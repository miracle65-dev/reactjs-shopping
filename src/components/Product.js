/**
 * @flow weak
 * @module Product
 * @extends React.Component
 *
 * @author Oleg Nosov <olegnosov1@gmail.com>
 * @license MIT
 *
 * @description
 * React component - Product form with price
 *
 */
import React, { Component, PropTypes } from 'react';
import animateScroll from 'react-scroll/lib/mixins/animate-scroll';
import capitalize from 'underscore.string/capitalize';

const
  /**
   * @static propTypes
   * @memberof Product
   *
   * @prop {string} name - Name to display. Required
   * @prop {string} path - Path to product. Required
   * @prop {number} price - Price (value only). Required
   * @prop {string} imagePath - Path to main image. Required
   * @prop {string} currency - Price currency. Default is £
   * @prop {Object.<string, Array<(string | number)>>} options - Custom product properties.
   * Default is {}
   * @prop {string} iconAddProductClassName - ClassName for cart icon on add to button.
   * Default is 'icon-cart-plus'
   */
  propTypes = {
    name: PropTypes.string.isRequired,
    path: PropTypes.string.isRequired,
    price: PropTypes.number.isRequired,
    imagePath: PropTypes.string.isRequired,
    currency: PropTypes.string,
    options: PropTypes.objectOf(
      PropTypes.arrayOf(
        PropTypes.oneOfType(
          [
            PropTypes.string,
            PropTypes.number,
          ]
        )
      )
    ),
    iconAddProductClassName: PropTypes.string,
  },
  /**
  * @static containerPropTypes
  * @memberof Product
  *
  * @prop {Function<string>} onAddProduct - Callback to call when
  * user want to add product in his cart
  * Example: onAddProduct('macbook-case', { quantity: 30, properties: { colour: 'red' } });
  * Required.
  * @prop {React Element} CheckoutButton - Button in the bottom of product.
  * Default is 'icon-cart-plus'
  */
  containerPropTypes = {
    onAddProduct: PropTypes.func.isRequired,
    CheckoutButton: PropTypes.element.isRequired,
  },
  defaultProps = {
    currency: '£',
    options: {},
    iconAddProductClassName: 'icon-cart-plus',
  };

function getAbsoluteOffsetTop({ offsetTop, offsetParent } : HTMLElement | Object = {}) : number {
  return offsetTop + (offsetParent && getAbsoluteOffsetTop(offsetParent));
}

export default class Product extends Component {

  static propTypes = { ...propTypes, ...containerPropTypes };
  static defaultProps = defaultProps;

  state = {
    quantity: 1,
  };

  handleQuantityValueChange = (
    { target: { value } } : { target : HTMLInputElement, }
  ) => {
    +value > -1 ? this.setState({ quantity: +value }) : 0;
  }

  handleSelectInputValueChange = (
    name : string,                 // product property name
    list : Array<string | number>, // product property options list
    { target: { value: listOptionName } } : { target : HTMLInputElement, }
  ) => {
    this.setState({ [name]: list.indexOf(listOptionName) });
  }

  /*
   * Return form-group consits of select input element
   * label value will be capitalized name
   * options - the list array
   */
  createSelectInputFromArray(name : string, list : Array<string | number>) : React$Element<any> {
    return (
      <div key={ name } className="form-group row">
        <label htmlFor={ name } className="col-xs-3 col-sm-3 col-md-3 col-lg-3 col-form-label">
          { capitalize(name) + ':' }
        </label>
        <div className="col-xs-9 col-sm-9 col-md-9 col-lg-9">
          <select
            onChange={ this.handleSelectInputValueChange.bind(this, name, list) }
            className="form-control"
            value={ list[this.state[name] || 0] }
          >
            {
              /*
               * Generate select input options based on list values
               */
              list.map((v, k) =>
                <option key={ v + k } value={ v }>
                  { capitalize(v) }
                </option>
              )
            }
          </select>
        </div>
      </div>
    );
  }

  generateProductProps() : Object {
    const { options, price, name, imagePath, currency } = this.props;
    const { quantity } = this.state;
    return (
      {
        quantity,
        properties: Object.entries(options).reduce((obj, [propName, list]) =>
        ({
          ...obj,
          [propName]: list[this.state[propName] || 0],
        })
      , {}),
        productInfo: {
          name,
          price,
          currency,
          imagePath,
        },
      }
    );
  }

  addProductFormSubmit = (event : Event) => {
    event.preventDefault();
    const { path, onAddProduct } = this.props;
    const { quantity } = this.state;
    setTimeout(target =>
      animateScroll.scrollTo(
        getAbsoluteOffsetTop(
          target.children[target.children.length - 2]) - 5, {
            duration: 750,
            delay: 0,
            smooth: true,
          })
    , 50, event.target);
    return quantity
           &&
           onAddProduct(
             path,
             this.generateProductProps()
           );
  }

  render() {
    const { price, currency, options, CheckoutButton, iconAddProductClassName } = this.props;

    return (
      <div>
        <p>
          <strong>
            Price: { currency + price.toFixed(2) }
          </strong>
        </p>
        <form className="m-t-1" onSubmit={ this.addProductFormSubmit }>
          {
            Object.entries(options).map(
              option => this.createSelectInputFromArray(...option)
            )
          }
          <div className="form-group row">
            <label htmlFor="product-quantity" className="col-xs-3 col-sm-3 col-md-3 col-lg-3 col-form-label">
              Quantity:
            </label>
            <div className="col-xs-9 col-sm-9 col-md-9 col-lg-9">
              <input onChange={ this.handleQuantityValueChange } className="form-control" type="number" value={ this.state.quantity } />
            </div>
          </div>
          <button type="submit" className="btn btn-success btn-block" disabled={ !this.state.quantity }>
            <i className={ iconAddProductClassName } />
            Add to cart
          </button>
          { CheckoutButton }
        </form>
      </div>
    );
  }
}
