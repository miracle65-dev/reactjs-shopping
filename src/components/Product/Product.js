/**
 * @flow
 * @module Product
 * @extends React.Component
 *
 * @author Oleg Nosov <olegnosov1@gmail.com>
 * @license MIT
 *
 * @description
 * React component - Product form with price.
 */
import React, { Component, PropTypes } from 'react';
import animateScroll from 'react-scroll/lib/mixins/animate-scroll';

import ProductPropertyInput from './ProductPropertyInput/ProductPropertyInput';
import {
  isNaturalNumber,
  getAbsoluteOffsetTop,
} from '../../helpers';

const
  /**
   * @static propTypes
   * @memberof Product
   *
   * @prop {string} name - Name to display. Required
   * @prop {id} id - Product's id. Required
   * @prop {string} path - Path to product. Required
   * @prop {Object.<string, number>} prices - Prices (currency-value). Required
   * @prop {string} imagePath - Path to main image. Required
   * @prop {string} currency - Price currency. Required
   * @prop {Object.<string, Array<(ProductPropertyOptionType)>>}
   * properties - Custom product properties. May be array of number, string or
   * shape({ additionalCost(optional), onSelect(optional), value(required)})
   * Default is {}
   * @prop {string} iconAddProductClassName - ClassName for cart icon
   * on add to button.
   * Default is 'icon-cart-plus'
   */
  propTypes = {
    name: PropTypes.string.isRequired,
    id: PropTypes.string.isRequired,
    path: PropTypes.string.isRequired,
    prices: PropTypes.objectOf(
      PropTypes.number,
    ).isRequired,
    imagePath: PropTypes.string.isRequired,
    currency: PropTypes.string.isRequired,
    properties: PropTypes.objectOf(
      PropTypes.arrayOf(
        PropTypes.oneOfType(
          [
            PropTypes.string,
            PropTypes.number,
            PropTypes.shape({
              additionalCost: PropTypes.objectOf(
                PropTypes.number,
              ),
              onSelect: PropTypes.func,
              value: PropTypes.oneOfType(
                [
                  PropTypes.string,
                  PropTypes.number,
                ],
              ).isRequired,
            }),
          ],
        ),
      ),
    ),
    propertiesToShowInCart: PropTypes.arrayOf(
      PropTypes.string,
    ),
    scrollAnimationConfig: PropTypes.object,
    iconAddProductClassName: PropTypes.string,
    afterPriceNode: PropTypes.node,
  },
  /**
  * @static containerPropTypes
  * @memberof Product
  * @prop {ReactElement} CheckoutButton - Button in the bottom of product.
  * Required.
  * @prop {Function<string>} onAddProduct - Callback to call when
  * user want to add product in his cart
  * Example:
  * onAddProduct(
  *   'macbook-case',
  *   { quantity: 30, properties: { colour: 'red' } },
  *   'GBP'
  * );
  * Required.
  * @prop {getLocalizationType} getLocalization - Required.
  * @prop {generateProductKey} generateProductKey - Function which generates
  * product's key based on id and properties. Example:
  * generateProductKey('macbook-case', { colour: 'red'} ).
  */
  containerPropTypes = {
    CheckoutButton: PropTypes.element.isRequired,
    onAddProduct: PropTypes.func.isRequired,
    getLocalization: PropTypes.func.isRequired,
    generateProductKey: PropTypes.func.isRequired,
  },
  defaultProps = {
    properties: {},
    propertiesToShowInCart: [],
    iconAddProductClassName: 'icon-cart-plus',
    scrollAnimationConfig: {
      duration: 750,
      delay: 0,
      smooth: true,
    },
    afterPriceNode: null,
  };

export default class Product extends Component {

  static propTypes = { ...propTypes, ...containerPropTypes };
  static defaultProps = defaultProps;

  static createPropertiesInputList = (
    properties: {
      [propertyName: string]: Array<ProductPropertyOptionType>
    },
    propertiesSelectedIndexes,
    currency,
    handlePropertyValueChange,
    getLocalization,
  ) : Array<React$Element<any>> =>
    Object
      .entries(properties)
      .map(([name, options]) =>
        <ProductPropertyInput
          key={name}
          name={name}
          options={options}
          selectedOptionIndex={propertiesSelectedIndexes[name]}
          currency={currency}
          onChange={handlePropertyValueChange}
          getLocalization={getLocalization}
        />,
    );

  static calculateAdditionalCost(
    properties : { [propName : string] : ProductPropertyOptionType },
    selectedPropertyIndexes : {[propName: string] : number},
    currency : string,
  ) : number {
    return Object.entries(properties).reduce(
      (sum, [propertyName, propertyOptions]) => {
        const selectedOption
          = propertyOptions[selectedPropertyIndexes[propertyName]|0];
        return sum + (
          typeof selectedOption === 'object'
          && (
            selectedOption.additionalCost
            && selectedOption.additionalCost[currency]
          ) || 0
        );
      }
    , 0);
  }

  static generateCartProduct(
    {
      properties,
      propertiesToShowInCart,
      prices,
      name,
      imagePath,
      path,
      id,
    } : {
      properties : { [propName : string] : ProductPropertyOptionType },
      propertiesToShowInCart : Array<string>,
      prices : { [currency : string] : number },
      name : string,
      imagePath : string,
      path : string,
      id : string,
    },
    quantity,
    selectedPropertyIndexes : {[propName: string] : number},
  ) : ProductType {
    const { calculateAdditionalCost } = Product;
    return {
      id,
      quantity,
      properties:
        Object
          .entries(properties)
          .reduce((obj, [propName, options]) =>
            ({
              ...obj,
              [propName]:
                ProductPropertyInput
                  .getOptionValue(
                    options[selectedPropertyIndexes[propName]|0],
                  ),
            })
          , {}),
      productInfo: {
        name,
        prices:
          Object
            .entries(prices)
            .reduce(
              (acc, [currency, price]) => ({
                ...acc,
                [currency]: price + calculateAdditionalCost(
                  properties,
                  selectedPropertyIndexes,
                  currency,
                ),
              }), {},
            ),
        path,
        imagePath,
        propertiesToShowInCart,
      },
    };
  }

  state = {
    quantity: 1,
  };

  componentWillUnmount() {
    this.clearScrollTimeout();
  }

  // Explicit define property for flow
  scrollTimeout = void 0;

  clearScrollTimeout = () => {
    if (typeof this.scrollTimeout !== 'undefined') {
      clearTimeout(this.scrollTimeout);
      delete this.scrollTimeout;
    }
  };

  handleQuantityValueChange = (
    { target: { value } } : { target : HTMLInputElement, },
  ) => {
    const quantity = +value;
    if (isNaturalNumber(quantity))
      this.setState({ quantity });
  }

  hanglePropertyValueChange = (
    { value }: { value: { [propName: string]: ProductPropertyOptionType }},
  ) => void this.setState(value);

  addProductFormSubmit = (event : Event) => {
    const { props } = this;
    const {
      id,
      path,
      currency,
      scrollAnimationConfig,
      onAddProduct,
      generateProductKey,
    } = props;
    const { quantity, ...selectedPropertyIndexes } = this.state;
    const { generateCartProduct } = Product;
    const { target: { children } } = event;
    event.preventDefault();

    if (quantity) {
      this.clearScrollTimeout();
      this.scrollTimeout = setTimeout(() =>
        void animateScroll.scrollTo(
          getAbsoluteOffsetTop(
            children[children.length - 2],
          ) - 5, scrollAnimationConfig,
        )
      , 50);
      const product = generateCartProduct(
        props, quantity, selectedPropertyIndexes,
      );
      onAddProduct(
         generateProductKey(
           id,
           product.properties,
         ),
         product,
         currency,
      );
    }
  }

  render() {
    const {
      addProductFormSubmit,
      handleQuantityValueChange,
      hanglePropertyValueChange,
      state,
      props,
    } = this;

    const {
      name,
      prices,
      currency,
      properties,
      iconAddProductClassName,
      afterPriceNode,
      CheckoutButton,
      getLocalization,
    } = props;

    const {
      quantity,
      ...selectedPropertyIndexes
    } = state;

    const {
      createPropertiesInputList,
      calculateAdditionalCost,
    } = Product;

    const price =
      prices[currency]
      + calculateAdditionalCost(
        properties,
        selectedPropertyIndexes,
        currency,
      );

    const localizationScope = {
      name,
      quantity,
      price,
      currency,
      get localizedCurrency() {
        return getLocalization(currency, localizationScope);
      },
      get localizedName() {
        return getLocalization(name, localizationScope);
      },
    };

    return (
      <div>
        <p>
          { getLocalization('price', localizationScope) }
        </p>
        { afterPriceNode }
        <form className="m-t-1" onSubmit={addProductFormSubmit}>
          {
            createPropertiesInputList(
              properties,
              selectedPropertyIndexes,
              currency,
              hanglePropertyValueChange,
              getLocalization,
            )
          }
          <div className="form-group row">
            <label
              htmlFor="product-quantity"
              className="col-xs-3 col-sm-3 col-md-3 col-lg-3 col-form-label"
            >
              { getLocalization('quantityLabel', localizationScope) }
            </label>
            <div className="col-xs-9 col-sm-9 col-md-9 col-lg-9">
              <input
                onChange={handleQuantityValueChange}
                className="form-control"
                type="number"
                value={quantity}
              />
            </div>
          </div>
          <button
            type="submit"
            className="btn btn-success btn-block"
            disabled={!quantity}
          >
            <i className={iconAddProductClassName} />
            {
              getLocalization('addToCart', localizationScope)
            }
          </button>
          { CheckoutButton }
        </form>
      </div>
    );
  }
}
