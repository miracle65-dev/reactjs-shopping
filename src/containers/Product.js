/**
 * @flow
 * @module ProductContainer
 *
 * @author Oleg Nosov <olegnosov1@gmail.com>
 * @license MIT
 *
 * @description
 * Redux container for Product
 */
import React from 'react';
import { connect } from 'react-redux';

import { Product } from '../components';
import CheckoutButton from '../containers/CheckoutButton';
import { addToCart } from '../actions';
import { configure, generateProductKey } from '../helpers';
import { getDefaultLocalization } from '../localization';

const mapStateToProps = (state : Object) : Object => ({
  currency: state.cart.currency,
});

const mapDispatchToProps = (dispatch : Function) : Object => ({
  onAddProduct: (key : string, product : Object, productCurrency: string) =>
    void dispatch(addToCart(key, product, productCurrency)),
});

export default(
  connect(mapStateToProps, mapDispatchToProps)(
    configure(Product, {
      checkoutButton: <CheckoutButton />,
      getLocalization: getDefaultLocalization('product'),
      generateProductKey,
    }),
  )
);
