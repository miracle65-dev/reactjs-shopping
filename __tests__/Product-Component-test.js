/*
 * Product tests for JEST
 *
 * Copyright © Oleg Nosov 2016
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE.txt file in the root directory of this source tree.
 */

import React from 'react';
import { mount } from 'enzyme';
import Product from '../src/components/Product';

const createProduct = ({ cartState, props }) =>
  mount(
    <Product
      { ...props }
      onAddProduct={
        (id, { quantity, productInfo, properties }) => { cartState[id + Object.values(properties).join('_')] = { quantity, id, ...productInfo, properties }; }
      }
      CheckoutButton={ <a /> }
    />);


describe('Product', () => {
  const iPadCaseInCart = {
    id: '/shop/ipad-case/',
    properties: {
      colour: 'red',
    },
    quantity: 1,
    productInfo: {
      name: 'iPad / Tablet case',
      price: 70,
      imagePath: '/shop/ipad-case/1-483x321.jpeg',
      currency: '£',
    },
  };

  const iPadCaseProps = {
    options: {
      colour: ['red', 'green'],
    },
    name: 'iPad / Tablet case',
    price: 70,
    path: '/shop/ipad-case/',
    imagePath: '1-483x321.jpeg',
  };

  it('adds products to cart', () => {
    // Create empty cart
    const cartState = {};

    // Current product is ipad case (see props above)
    const renderedProduct = createProduct({ cartState, props: iPadCaseProps });

    const simulateAddProductEvent = () => renderedProduct.find('form').simulate('submit');

    simulateAddProductEvent();


    // Our product is in cart already
    expect(cartState['/shop/ipad-case/red'].quantity).toBe(1);

    // Try to change quantity to -1
    renderedProduct.find('input').simulate('change', { target: { value: -1 } });


    expect(renderedProduct.state().quantity).toBe(1);

    // Now add green case in our cart

    renderedProduct.find('select').simulate('change', { target: { value: 'green' } });

    simulateAddProductEvent();

    // Have two products in cart
    expect(Object.keys(cartState).length).toBe(2);
  });

  it('takes snapshot', () => {
    const cartState = {};
    const renderedProduct = createProduct({ cartState, props: iPadCaseProps });
    expect(renderedProduct.html()).toMatchSnapshot();
  });
});
