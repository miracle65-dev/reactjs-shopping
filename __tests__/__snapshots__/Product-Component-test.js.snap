exports[`Product takes snapshot 1`] = `
<div>
  <p>
    Price: Â£70
  </p>
  <form
    className="mt-1"
    onSubmit={[Function anonymous]}>
    <div
      className="form-group row">
      <label
        className="col-xs-3 col-sm-3 col-md-3 col-lg-3 col-form-label"
        htmlFor="colour">
        Colour:
      </label>
      <div
        className="col-xs-9 col-sm-9 col-md-9 col-lg-9">
        <select
          className="form-control"
          onChange={[Function anonymous]}
          value="red">
          <option
            value="red">
            Red
          </option>
          <option
            value="green">
            Green
          </option>
        </select>
      </div>
    </div>
    <div
      className="form-group row">
      <label
        className="col-xs-3 col-sm-3 col-md-3 col-lg-3 col-form-label"
        htmlFor="product-quantity">
        Quantity:
      </label>
      <div
        className="col-xs-9 col-sm-9 col-md-9 col-lg-9">
        <input
          className="form-control"
          onChange={[Function anonymous]}
          type="number"
          value={1} />
      </div>
    </div>
    <button
      className="btn btn-success btn-block"
      disabled={false}
      type="submit">
      <i
        className="icon-cart-plus" />
      Add to my cart :)
    </button>
  </form>
</div>
`;
